lib/Parse/ParseExpr.cpp

```diff
@@ -3007,6 +3016,107 @@ bool Parser::ParseRelativeBoundsClause(ExprResult &Expr) {
   return IsError;
 }
 
+ExprResult Parser::ParseBoundsCastExpression(IdentifierInfo &Ident,
+                                             SourceLocation &ILoc) {
+  ExprResult Result(true);
+  SourceLocation StartLoc = Tok.getLocation();
+  ParsedType CastTy = nullptr;
+  BoundsCastExpr::Kind kind;
+  if (&Ident == Ident_dynamic_bounds_cast)
+    kind = BoundsCastExpr::Kind::Dynamic;
+  else if (&Ident == Ident_assume_bounds_cast)
+    kind = BoundsCastExpr::Kind::Assume;
+  else
+    kind = BoundsCastExpr::Kind::Invalid;
+
+  DeclSpec DS(AttrFactory);
+  DS.SetRangeStart(Tok.getLocation());
+  DS.SetRangeEnd(SourceLocation());
+
+  if (ExpectAndConsume(tok::less, diag::err_expected_less_after,
+                       Ident.getNameStart()))
+    return ExprError();
+
+  TypeResult Ty = ParseTypeName();
+  if (Ty.isInvalid())
+    SkipUntil(tok::greater, StopAtSemi);
+
+  // if there is rel, parsing it.
+  // if(Tok.is(tok::comma)){
+  //   ConsumeToken();
+  //   Parsing Alignment
+  //   .....
+  // }
+
+  Declarator DeclaratorInfo(DS, Declarator::TypeNameContext);
```

Comment:
I don't think you should be parsing a declarator here.   ParseTypeName already parses a declarator.

You can look at ParseInteropTypeAnnotation (line 2780, ParseExpr.cpp) to see an example of how to ParseTypeName can be used.